src/Common/threadPoolCallbackRunner.h

```diff
@@ -123,13 +128,17 @@ class ThreadPoolCallbackRunnerLocal final
     template <typename Function>
     static void executeCallback(std::promise<void> & promise, Function & callback)
     {
+        /// Release callback before setting value to the promise to avoid
+        /// destruction of captured resources after waitForAllToFinish returns.
         try
         {
             callback();
+            callback = {};
```

Comment:
why not have `Function && callback`?
so you move callback when calling `executeCallback` and you don't have to manually destroy it `operator()`

Reply:
I want to destroy the callback before the promise is set. Otherwise, the behavior will be almost identical to the current one.

Reply:
it's more of a nitpick just to make ti clear that callback is unusable
but it's really small and doesn't affect behaviour so let's just merge the fix ASAP